frame/contracts/src/lib.rs

```diff
@@ -563,31 +549,74 @@ decl_module! {
 			let mut gas_meter = GasMeter::new(gas_limit);
-
 			let result = Self::execute_wasm(origin, &mut gas_meter, |ctx, gas_meter| {
 				ctx.call(dest, value, gas_meter, data)
 			});
-			gas_meter.into_dispatch_result(result)
+			gas_meter.into_dispatch_result(result, T::WeightInfo::call())
 		}
 
-		/// Instantiates a new contract from the `code_hash` generated by `put_code`,
-		/// optionally transferring some balance.
+		/// Instantiates a new contract from the supplied `code` optionally transferring
+		/// some balance.
+		///
+		/// This is the only function that can deploy new code to the chain.
 		///
-		/// The supplied `salt` is used for contract address deriviation. See `fn contract_address`.
+		/// # Parameters
+		///
+		/// * `endowment`: The balance to transfer from the `origin` to the newly created contract.
+		/// * `gas_limit`: The gas limit enforced when executing the constructor.
+		/// * `code`: The contract code to deploy in raw bytes.
+		/// * `data`: The input data to pass to the contract constructor.
+		/// * `salt`: Used for the address derivation. See [`Self::contract_address`].
 		///
 		/// Instantiation is executed as follows:
 		///
+		/// - The supplied `code` is instrumented, deployed, and a `code_hash` is created for that code.
+		/// - If the `code_hash` already exists on the chain the underlying `code` will be shared.
 		/// - The destination address is computed based on the sender, code_hash and the salt.
 		/// - The smart-contract account is created at the computed address.
-		/// - The `ctor_code` is executed in the context of the newly-created account. Buffer returned
-		///   after the execution is saved as the `code` of the account. That code will be invoked
-		///   upon any call received by this account.
-		/// - The contract is initialized.
+		/// - The `endowment` is transferred to the new account.
+		/// - The `deploy` function is executed in the context of the newly-created account.
 		#[weight =
-			T::WeightInfo::instantiate(
-				data.len() as u32 / 1024,
+			T::WeightInfo::instantiate_with_code(
+				code.len() as u32 / 1024,
 				salt.len() as u32 / 1024,
-			).saturating_add(*gas_limit)
+			)
+			.saturating_add(*gas_limit)
+		]
+		pub fn instantiate_with_code(
+			origin,
+			#[compact] endowment: BalanceOf<T>,
+			#[compact] gas_limit: Gas,
+			code: Vec<u8>,
+			data: Vec<u8>,
+			salt: Vec<u8>,
+		) -> DispatchResultWithPostInfo {
+			let origin = ensure_signed(origin)?;
+			let schedule = <Module<T>>::current_schedule();
+			let code_len = code.len() as u32;
+			ensure!(code_len <= schedule.limits.code_size, Error::<T>::CodeTooLarge);
+			let mut gas_meter = GasMeter::new(gas_limit);
+			let result = Self::execute_wasm(origin, &mut gas_meter, |ctx, gas_meter| {
+				let executable = PrefabWasmModule::from_code(code, &schedule)?;
+				let result = ctx.instantiate(endowment, gas_meter, &executable, data, &salt)
+					.map(|(_address, output)| output)?;
+				executable.store();
```

Comment:
same as above mentioned, should executable be stored before ctx.instantiate ?

Reply:
I went out of my way to have it in this specific order. The reason is that the instantiate can fail but the store cannot. Doing it this way around we avoid a storage transaction and possible rollback of the code storage.

Reply:
yes ok then